src/util/BlockingCache.hpp

```diff
@@ -0,0 +1,221 @@
+     * @brief Default constructor - creates an empty cache
+     */
+    BlockingCache() = default;
+
+    /**
+     * @brief Construct a cache with an initial value
+     * @param initialValue The value to initialize the cache with
+     */
+    explicit BlockingCache(ValueType initialValue) : state_{State::Full}, value_(std::move(initialValue))
+    {
+    }
+
+    BlockingCache(BlockingCache&&) = delete;
+    BlockingCache(BlockingCache const&) = delete;
+    BlockingCache&
+    operator=(BlockingCache&&) = delete;
+    BlockingCache&
+    operator=(BlockingCache const&) = delete;
+
+    /**
+     * @brief Function type for cache update operations
+     * @details Called when the cache needs to be populated or refreshed
+     */
+    using Updater = std::function<std::expected<ValueType, ErrorType>(boost::asio::yield_context)>;
+
+    /**
+     * @brief Function type to verify if a value should be cached
+     * @details Returns true if the value should be stored in the cache
+     */
+    using Verifier = std::function<bool(ValueType const&)>;
+
+    /**
+     * @brief Asynchronously get a value from the cache, updating if necessary
+     *
+     * @param yield The asio yield context for coroutine suspension
+     * @param updater Function to generate a new value if needed
+     * @param verifier Function to validate whether a value should be cached
+     * @return std::expected<ValueType, ErrorType> The cached value or an error
+     *
+     * Depending on the current cache state, this will either:
+     * - Return the cached value if it's already present
+     * - Wait for an ongoing update to complete
+     * - Trigger a new update if the cache is empty
+     */
+    [[nodiscard]] std::expected<ValueType, ErrorType>
+    asyncGet(boost::asio::yield_context yield, Updater updater, Verifier verifier)
+    {
+        switch (state_) {
+            case State::Updating: {
+                return wait(yield, std::move(updater), std::move(verifier));
+            }
+            case State::Full: {
+                auto const value = value_.template lock<std::shared_lock>();
+                ASSERT(value->has_value(), "Value should be presented when the cache is full");
+                return value->value();
+            }
+            case State::Empty: {
+                return update(yield, std::move(updater), std::move(verifier));
+            }
+        };
+        std::unreachable();
+    }
+
+    /**
+     * @brief Force an update of the cache value
+     *
+     * @param yield The ASIO yield context for coroutine suspension
+     * @param updater Function to generate a new value
+     * @param verifier Function to validate whether a value should be cached
+     * @return std::expected<ValueType, ErrorType> The new value or an error
+     *
+     * Initiates a cache update operation regardless of current state.
+     * If another update is already in progress, waits for it to complete.
+     */
+    std::expected<ValueType, ErrorType>
```

Comment:
Don't we want `nodiscard` here too or is it implied that the user may just want to update but does not care about the result?

Reply:
Good point, I will add `nodiscard`.